robocup_ws/src/Planner/src/player_controller.py

```diff
@@ -213,31 +314,55 @@ def ball_is_free(self,game_info: list) -> [bool, int, int]:
                 return False, 0, min_arg_opponents
         return True, -1, -1
 
-    def receive_ball(self, game_info, partner_id):
-
-        team = game_info[0]
-        player_id = game_info[2]
-
-        partner = team.players_positions_efcs[partner_id]
-
-        pass_target = team.players_positions_efcs[player_id]
-        ball_position = team.ball_pos_efcs
 
-        delta_position = np.array([partner.x, partner.y]) - np.array([pass_target.x, pass_target.y])
-        calculated_heading = np.arctan2(delta_position[1],delta_position[0])
-        if(calculated_heading is None):
-            if(delta_position[1] >= 0):
-                new_heading = np.pi/2
+    def receive_ball(self, robot_state: Position, target: Position, ball_position: Position, ball_vel: Position, my_robot_id, player_to_pass):
+        d = np.hypot(robot_state.x - ball_position.x, robot_state.y - ball_position.y)
+        action = 0
+        if d < 0.17:
+            direct_kick_feasible, kick_x, kick_y = TeamMasterSupporting.check_if_direct_goal_feasible(None,
+                                                                                                      self.team_position.ball_pos_efcs,
+                                                                                                      self.opponents_position.players_positions_wcs,
+                                                                                                      self.team_id)
+            if direct_kick_feasible and my_robot_id in [self.striker_right_idx, self.striker_left_idx]:
+                temp_target_pos = Position(kick_x, kick_y, 0)
+                raw_command = receive_and_pass_action(robot_state, temp_target_pos, ball_position, ball_vel)
+                command = PlayerCommand(*raw_command)
             else:
-                new_heading = -np.pi/2
+                temp_target_pos = self.team_position.players_positions_efcs[player_to_pass]
+                raw_command = receive_and_pass_action(robot_state, temp_target_pos, ball_position, ball_vel)
+                command = PlayerCommand(*raw_command)
         else:
-            new_heading = calculated_heading
-
-        vel_l, vel_r, action, _ = rotate_towards(pass_target, new_heading)
+            command = PlayerCommand(*go_to_fast(robot_state, target), action)
+        return command
 
-        return PlayerCommand(vel_l, vel_r, action)
+    def avoid_obstacle(self, game_info: list, team_id, obstacle_player_id) -> PlayerCommand:
+        team = game_info[0]
+        obstacle_team = game_info[team_id]
+        player_id = game_info[2]
+        my_pos = team.players_positions_efcs[player_id]
+        safe_pos = obstacle_team.players_positions_wcs[obstacle_player_id]
 
+        lv, rv = self.go_around_the_point(my_pos, safe_pos, 0.3)
+        return PlayerCommand(lv, rv, 0)
 
+    def go_around_the_point(self, robot_state: Position, go_around_point: Position, radius, direction=1, R_vel=3):
```

Comment:
this seems like copy-paste from /go_to_action.py, even if you need some small changes modify it there

If sb found a bug in this code it is enough to change it in one place - otherwise, it will get messy and get multiple versions of the same functions with different bug fixes